# Fase 2 Limpieza de los datos

Tras unir la tabla es momento de unir datos, vamos a seguir reduciendo nulos y a eliminar columnas no deseadas

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import sys
sys.path.append("../") #Para poder importar
from src.SupportCleaning import fill_organo, fill_unidad_gestora,fill_missing_organo_superior,fill_missing_organo
warnings.filterwarnings('ignore')
pd.options.display.float_format = '{:.2f}'.format
# Carga el dataset de trabajo
pd.set_option("display.max_columns", None) #Ver todo
df = pd.read_csv("../datos/output/datos-2013_to_2021Fase1.csv", delimiter=",",index_col=0)


# Objetivo: Reducir nulos a 0

Si es posible, ahora que hemos unido los datos, vamos a intentar sacar nuevos diccionarios que nos permita dejar todo rellenado correctamente sin nulos

In [2]:
condicion = df["código del órgano superior"].isnull()
df2 = df[condicion]
df2["nombre del órgano superior"].unique()

array([nan], dtype=object)

In [3]:
condicion = df["código del órgano"].isnull()
df2 = df[condicion]
df2["nombre del órgano"].unique()

array([nan, 'Companhia Docas do Rio Grande do Norte',
       'Controladoria-Geral da União - Unidades com vínculo direto',
       'Secretaria Especial de Agricultura Familiar e do Desenvolvimento Agrário',
       'Amazônia Azul Tecnologia de Defesa S.A.',
       'Agência Nacional de Mineração',
       'UNIVERSIDADE FEDERAL DO DELTA DO PARNAIBA',
       'TELECOMUNICACOES BRASILEIRAS S/A'], dtype=object)

In [4]:
condicion = df["código de la unidad gestora"].isnull()
df2 = df[condicion]
df2["nombre de la unidad gestora"].unique()

array([nan, 'SECRETARIA EXECUTIVA-MPA', 'SECRETARIA DE PORTOS/PR',
       'SECRETARIA DE AVIACAO CIVIL/PR',
       'FUNDO NAC.P/A CRIANCA E O ADOLESC.SET.ORC/FIN',
       'INST.FED. FLUMINENSE/CAMPUS CAMPOS-CENTRO',
       'SECRETARIA ESPECIAL DE AQUICULTURA E PESCA/PR',
       'PROCURADORIA GERAL DA FAZENDA NACIONAL',
       'SETORIAL ORCAMENTARIA E FINANCEIRA - ANM',
       'COORDENACAO-GERAL DO FUNDO DE TERRAS',
       'FDO EXERCITO - DIFERENCA CAMBIAL',
       'UNIVERSIDADE FEDERAL DO DELTA DO PARNAIBA',
       'COORDENACAO DE ORCAMENTO E FINANCAS - PRF',
       'TELECOMUNICACOES BRASILEIRAS S/A',
       'FUNDO NACIONAL DA CULTURA - FNC', 'UNIVERSIDADE FEDERAL DE JATAI',
       'UNIVERSIDADE FEDERAL DE RONDONOPOLIS',
       'UNIVERSIDADE FEDERAL DE CATALAO', 'MRE - DIFERENCA CAMBIAL'],
      dtype=object)

Vemos que todos los códigos y nombres vacios en las 2 primeras columnas se produce porque no hay ninguno de los dos datos rellenados, pero en organo y unidad gestora, vemos que no es así vamos a empezar por ahí.

Para corregirlo, vamos a generar un nuevo diccionario, pero esta vez usando los nombres como índices


In [5]:
datos = df.groupby("código del órgano")["nombre del órgano"].first()
lista_codigos = list(datos.index)
lista_names = list(datos.values)
diccionario_organos = {}
for i in range(0, len(lista_codigos)):
    diccionario_organos[lista_codigos[i]] = lista_names[i]  

diccionario_organos 

{20101.0: 'Presidência da República',
 20116.0: 'Fundo de Imprensa Nacional',
 20117.0: 'Fundo Nacional Antidrogas',
 20122.0: 'Secretaria Nacional de Políticas para Mulheres',
 20129.0: 'Secretaria Especial de Agricultura Familiar e do Desenvolvimento Agrário',
 20202.0: 'Fundação Escola Nacional de Administração Pública',
 20203.0: 'Agência Nacional do Cinema',
 20301.0: 'Comissão Nacional de Energia Nuclear',
 20302.0: 'Nuclebrás Equipamentos Pesados S.A.',
 20303.0: 'Indústrias Nucleares do Brasil',
 20402.0: 'Agência Espacial Brasileira',
 20403.0: 'Fundação Casa de Rui Barbosa',
 20404.0: 'Fundação Biblioteca Nacional',
 20408.0: 'Fundação Cultural Palmares',
 20411.0: 'Instituto do Patrimônio Histórico e Artístico Nacional',
 20412.0: 'Fundação Nacional de Artes',
 20415.0: 'Empresa Brasil de Comunicação',
 20501.0: 'Conselho Nacional de Desenvolvimento Científico e Tecnológico',
 20502.0: 'Financiadora de Estudos e Projetos',
 20603.0: 'Superintendência da Zona Franca de Manaus

Veamos los nulos antes de llamar a la función

In [6]:
nulos_codorg = df["código del órgano"].isnull().sum()
nulos_nameorg = df["nombre del órgano"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

1186
1302


Llamamos a la función que hemos creado para reemplazar nulos por valores válidos

In [7]:
fill_organo(df,diccionario_organos) #Pasamos el diccionario, para que si algún día se extiende 

In [8]:
nulos_codorg = df["código del órgano"].isnull().sum()
nulos_nameorg = df["nombre del órgano"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

905
905


In [9]:
condicion = df["código del órgano"].isnull()
df2 = df[condicion]
df2["nombre del órgano"].unique()

array([nan], dtype=object)

Ya hemos dejado los organos con nulos sólo si nombre y código es nulo, hagamos lo mismo con unidad gestora

In [10]:
datos = df.groupby("código de la unidad gestora")["nombre de la unidad gestora"].first()
lista_codigos2 = list(datos.index)
lista_names2 = list(datos.values)
diccionario2 = {}
for i in range(0, len(lista_codigos2)):
    diccionario2[lista_codigos2[i]] = lista_names2[i]    
    


In [11]:
nulos_codorg = df["código de la unidad gestora"].isnull().sum()
nulos_nameorg = df["nombre de la unidad gestora"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

1053
947


In [12]:
fill_unidad_gestora(df,diccionario2)

In [13]:
nulos_codorg = df["código de la unidad gestora"].isnull().sum()
nulos_nameorg = df["nombre de la unidad gestora"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

614
614


# Convertir los nulos en valores válidos

Ahora que ya tenemos esta parte, vamos a intentar convertir en valores válidos los nombres y códigos nulos de todas las columnas
Primero veamos si hay correlación 

In [14]:
condicion = df["código de la unidad gestora"].isnull()
df2 = df[condicion]
df2

código del órgano superior  \
74                         22000.00   
3831                       32000.00   
3855                       32000.00   
4423                       54000.00   
4427                       54000.00   
...                             ...   
1023953                    54000.00   
1025277                    54000.00   
1025568                    54000.00   
1025717                         NaN   
1026226                    20000.00   

                            nombre del órgano superior  código del órgano  \
74       Ministério da Agricultura, Pecuária e Abastec           22202.00   
3831                     Ministério de Minas e Energia           32205.00   
3855                     Ministério de Minas e Energia           32263.00   
4423                             Ministério do Turismo           20411.00   
4427                             Ministério do Turismo           20412.00   
...                                                ...                ...   
1023953                          Ministério do Turismo           34902.00   
1025277                          Ministério do Turismo           20404.00   
1025568                          Ministério do Turismo           20203.00   
1025717                                            NaN           20411.00   
1026226                       Presidência da República           20116.00   

                                         nombre del órgano  \
74             Empresa Brasileira de Pesquisa Agropecuária   
3831     Agência Nacional do Petróleo, Gás Natural e Bi...   
3855             Departamento Nacional de Produção Mineral   
4423     Instituto do Patrimônio Histórico e Artístico ...   
4427                            Fundação Nacional de Artes   
...                                                    ...   
1023953                          Fundo Nacional de Cultura   
1025277                       Fundação Biblioteca Nacional   
1025568                         Agência Nacional do Cinema   
1025717  Instituto do Patrimônio Histórico e Artístico ...   
1026226                         Fundo de Imprensa Nacional   

         código de la unidad gestora nombre de la unidad gestora  \
74                               NaN                         NaN   
3831                             NaN                         NaN   
3855                             NaN                         NaN   
4423                             NaN                         NaN   
4427                             NaN                         NaN   
...                              ...                         ...   
1023953                          NaN                         NaN   
1025277                          NaN                         NaN   
1025568                          NaN                         NaN   
1025717                          NaN                         NaN   
1026226                          NaN                         NaN   

                              categoría económica     origen de los ingresos  \
74                             Receitas Correntes  Outras Receitas Correntes   
3831                           Receitas Correntes  Outras Receitas Correntes   
3855                           Receitas Correntes  Outras Receitas Correntes   
4423                           Receitas Correntes        Receita de Serviços   
4427                           Receitas Correntes  Outras Receitas Correntes   
...                                           ...                        ...   
1023953                        Receitas Correntes        Receita Patrimonial   
1025277                        Receitas Correntes        Receita de Serviços   
1025568                        Receitas Correntes  Outras Receitas Correntes   
1025717                        Receitas Correntes        Receita de Serviços   
1026226  Receitas Correntes - intra-orçamentárias         Receita Industrial   

                                      tipo de ingresos  \
74                           Demai

In [15]:
df.groupby(["código del órgano", "código de la unidad gestora", "nombre de la unidad gestora"]).size().reset_index(name='counts').sort_values(by="código del órgano")



código del órgano  código de la unidad gestora  \
0             20101.00                    110005.00   
1             20116.00                    110245.00   
2             20117.00                    110246.00   
3             20122.00                    200021.00   
4             20129.00                    110690.00   
..                 ...                          ...   
363           74918.00                    533009.00   
364           74919.00                    537004.00   
365           81000.00                    810002.00   
366           90000.00                    390001.00   
367           91214.00                    913001.00   

                       nombre de la unidad gestora  counts  
0    SECRETARIA DE ADMINISTRACAO-SET.ORC.E FIN./PR    1548  
1      FUNDO DE IMPRENSA NACIONAL/EXEC.ORC.FINANC.    2937  
2                        FUNDO NACIONAL ANTIDROGAS      20  
3         SECRETARIA DE POLITICAS PARA AS MULHERES     166  
4     SEC.ESPEC.DE AGRIC.FAMIL.E DO DESENV.AGRARIO     472  
..                                             ...     ...  
363           FUNDO DE DESENVOLVIMENTO DO NORDESTE     502  
364       FUNDO DE DESENVOLVIMENTO DO CENTRO-OESTE     237  
365       SETORIAL ORCAMENTARIA E FINANCEIRA - MDH      15  
366                   RESERVA DE CONTINGENCIA/MEFP      25  
367                    AUTORIDADE PUBLICA OLIMPICA     147  

[368 rows x 4 columns]

In [16]:
c1 = df["código del órgano superior"] == 20000
cond = df["código del órgano"] == 67000 
a = df[c1]
a["código del órgano"].unique()

array([67000., 20101., 20116., 69000., 24208., 64000., 60000., 20117.,
       61000.,    nan])

In [17]:
c1 = df["código del órgano"] == 22202
cond = df["código del órgano"] == 67000 
a = df[c1]
a["código de la unidad gestora"].unique()

array([135037.,     nan])

In [18]:
c1 = df["código del órgano superior"] == 20000
cond = df["código del órgano"] == 20101 
a = df[cond]
a["código del órgano superior"].unique()

array([   nan, 20000.])

In [19]:
df2 = df[c1]
df2["código del órgano superior"].unique()

array([20000.])

In [20]:
diccionario_organo_unidad = df.groupby("código del órgano")["código de la unidad gestora"].first().dropna().to_dict()

In [21]:
cond = df["código del órgano superior"].isnull()
cond2 = df["código de la unidad gestora"].isnull()
df[cond & cond2]

código del órgano superior nombre del órgano superior  \
150126                          NaN                        NaN   
275212                          NaN                        NaN   
525593                          NaN                        NaN   
878054                          NaN                        NaN   
929868                          NaN                        NaN   
1025717                         NaN                        NaN   

         código del órgano                                  nombre del órgano  \
150126            26280.00        Fundação Universidade Federal de São Carlos   
275212            26263.00                     Universidade Federal de Lavras   
525593            26427.00                         Instituto Federal da Bahia   
878054            32396.00                      Agência Nacional de Mineração   
929868            25000.00  Ministério da Economia - Unidades com vínculo ...   
1025717           20411.00  Instituto do Patrimônio Histórico e Artístico ...   

         código de la unidad gestora nombre de la unidad gestora  \
150126                           NaN                         NaN   
275212                           NaN                         NaN   
525593                           NaN                         NaN   
878054                           NaN                         NaN   
929868                           NaN                         NaN   
1025717                          NaN                         NaN   

        categoría económica origen de los ingresos  \
150126   Receitas Correntes    Receita de Serviços   
275212   Receitas Correntes    Receita de Serviços   
525593   Receitas Correntes    Receita Patrimonial   
878054   Receitas Correntes    Receita de Serviços   
929868   Receitas Correntes          Contribuições   
1025717  Receitas Correntes    Receita de Serviços   

                                     tipo de ingresos  \
150126   Serviços Administrativos e Comerciais Gerais   
275212   Serviços Administrativos e Comerciais Gerais   
525593                            Valores Mobiliários   
878054   Serviços Administrativos e Comerciais Gerais   
929868                       Contribuições econômicas   
1025717  Serviços Administrativos e Comerciais Gerais   

                                               detalle  \
150126   INSCR.EM CONCURSOS E PROC.SELETIVOS-PRINCIPAL   
275212   SERVICOS DE INFORMACAO E TECNOLOGIA-PRINCIPAL   
525593    REMUNERACAO DE DEPOSITOS BANCARIOS-PRINCIPAL   
878054     SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   
929868   CONTR.S/REC.BRT EMP.PREST.SERV.TELECOM-PRINC.   
1025717    SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

         valor previsto actualizado  valor registrado  valor ejecutado  \
150126                         0.00              0.00          1012.00   
275212                         0.00              0.00           427.50   
525593                         0.00              0.00            51.18   
878054                         0.00              0.00           626.39   
929868                         0.00              0.00          7921.96   
1025717                        0.00              0.00            74.00   

         porcentaje ejecutado fecha de registro  año fiscal  
150126                   0.00        2016-11-11        2016  
275212                   0.00        2017-02-17        2017  
525593                   0.00        2018-08-09        2018  
878054                   0.00        2020-02-20        2020  
929868                   0.00        2021-07-22        2021  
1025717                  0.00        2021-09-15        2021

# Que vamos a hacer
En funcion de lo anterior, nos fijamos que un código de órgano superior tiene múltiples códigos organo, si sacamos un diccionario de que códigos órgano apuntan a qué códigos organo superior podremos rellenar las columnas por código

Pero antes haremos lo mismo con el órgano y unidad gestora y ver si funciona

In [22]:
df.groupby("código de la unidad gestora")["código del órgano"].first().reset_index()

código de la unidad gestora  código del órgano
0                      110005.00           20101.00
1                      110008.00           58000.00
2                      110060.00           63000.00
3                      110101.00           60000.00
4                      110174.00           66000.00
..                           ...                ...
359                    773009.00           52233.00
360                    806001.00           25207.00
361                    810002.00           81000.00
362                    873001.00           52133.00
363                    913001.00           91214.00

[364 rows x 2 columns]

In [23]:
datos = df.groupby("código de la unidad gestora")["código del órgano"].first().sort_values()
lista_codigos4 = list(datos.index)
lista_names4 = list(datos.values)
diccionario4 = {}
for i in range(0, len(lista_codigos4)):
    diccionario4[lista_codigos4[i]] = lista_names4[i]    

datos = df.groupby("código del órgano")["nombre del órgano"].first()
lista_codigos = list(datos.index)
lista_names = list(datos.values)
diccionario_organos = {}
for i in range(0, len(lista_codigos)):
    diccionario_organos[lista_codigos[i]] = lista_names[i]  



In [24]:
nulos_codorg = df["código del órgano"].isnull().sum()
nulos_nameorg = df["nombre del órgano"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

905
905


In [25]:
fill_missing_organo(df,diccionario4,diccionario_organos)

In [26]:
nulos_codorg = df["código del órgano"].isnull().sum()
nulos_nameorg = df["nombre del órgano"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

0
0


In [27]:
c = df["nombre del órgano superior"].isnull()
df[c].head(1)

código del órgano superior nombre del órgano superior  código del órgano  \
31                         NaN                        NaN           22905.00   

                       nombre del órgano  código de la unidad gestora  \
31  Fundo de Defesa da Economia Cafeeira                    130137.00   

                     nombre de la unidad gestora categoría económica  \
31  FUNDO DE DEFESA DA ECON.CAFEEIRA/FUNCAFE/SPA  Receitas Correntes   

       origen de los ingresos                             tipo de ingresos  \
31  Outras Receitas Correntes  Indenizações, restituições e ressarcimentos   

                                         detalle  valor previsto actualizado  \
31  RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                    61810.00   

    valor registrado  valor ejecutado  porcentaje ejecutado fecha de registro  \
31              0.00             0.00                  0.00        2013-12-31   

    año fiscal  
31        2013

# Ha funcionado! 
Nos quedan 13 nulos que se debe a que estaban las 4 columnas en Null

Hagamos lo mismo para órgano superior con órgano

In [28]:
datos = df.groupby("código del órgano")["código del órgano superior"].first().sort_values()
lista_codigos3 = list(datos.index)
lista_names3 = list(datos.values)
diccionario3 = {}
for i in range(0, len(lista_codigos3)):
    diccionario3[lista_codigos3[i]] = lista_names3[i]    


In [29]:
nulos_codorg = df["código del órgano superior"].isnull().sum()
nulos_nameorg = df["nombre del órgano superior"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

10545
10545


In [30]:
fill_missing_organo_superior(df,diccionario3)

In [31]:
nulos_codorg = df["código del órgano superior"].isnull().sum()
nulos_nameorg = df["nombre del órgano superior"].isnull().sum()
print(nulos_codorg)
print(nulos_nameorg)

0
0


Ahora ya tenemos las 4 primeras columnas sin valores nulos, podemos convertir los datos a int

In [32]:
df["código del órgano superior"] = df["código del órgano superior"].astype(int)
df["código del órgano"] = df["código del órgano"].astype(int)

In [33]:
check_unidad_multiple_organos = df.groupby("código de la unidad gestora")["código del órgano"].nunique()
unidades_multiples_organos = check_unidad_multiple_organos[check_unidad_multiple_organos > 1]

# Mostrar las unidades gestoras que están vinculadas a más de un órgano (si existen)
print(unidades_multiples_organos)

código de la unidad gestora
307001.00    2
307002.00    2
410002.00    2
873001.00    2
Name: código del órgano, dtype: int64


In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   código del órgano superior   1026299 non-null  int64  
 1   nombre del órgano superior   1026299 non-null  object 
 2   código del órgano            1026299 non-null  int64  
 3   nombre del órgano            1026299 non-null  object 
 4   código de la unidad gestora  1025685 non-null  float64
 5   nombre de la unidad gestora  1025685 non-null  object 
 6   categoría económica          1007321 non-null  object 
 7   origen de los ingresos       987881 non-null   object 
 8   tipo de ingresos             994372 non-null   object 
 9   detalle                      996962 non-null   object 
 10  valor previsto actualizado   974984 non-null   float64
 11  valor registrado             999880 non-null   float64
 12  valor ejecutado              986834 non-null   

In [35]:
c = df["código de la unidad gestora"] == 873001
d2 = df[c]
d2.head(1)

código del órgano superior nombre del órgano superior  código del órgano  \
374                       52000       Ministério da Defesa              52133   

                                     nombre del órgano  \
374  Fundo de Desenvolvimento do Ensino Profissiona...   

     código de la unidad gestora  \
374                    873001.00   

                      nombre de la unidad gestora categoría económica  \
374  DIRETORIA DE FINANCAS-SISTEMA PARA O PAIS/MM  Receitas Correntes   

        origen de los ingresos                             tipo de ingresos  \
374  Outras Receitas Correntes  Indenizações, restituições e ressarcimentos   

                                          detalle  valor previsto actualizado  \
374  RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                        0.00   

     valor registrado  valor ejecutado  porcentaje ejecutado  \
374              0.00        658783.99                  0.00   

    fecha de registro  año fiscal  
374        2013-12-31        2013

# Clasificar facturas nulas como "sin información"

Para las columnas categoría económica, origen de los ingresos y tipo de ingresos vamos a cambiar los valores nulos por strings, vamos primero a comprobar como hacerlo



In [36]:
df["categoría económica"].isnull().sum()        #Sem informação


np.int64(18978)

In [37]:
df["origen de los ingresos"].isnull().sum() #Sem informação

np.int64(38418)

In [38]:
df["tipo de ingresos"].isnull().sum() # lo pasaremos a -> Sem informação

np.int64(31927)

In [39]:
# Reemplazar los valores nulos en las columnas especificadas por "Sem informação"
df["tipo de ingresos"] = df["tipo de ingresos"].fillna("Sem informação")
df["origen de los ingresos"] = df["origen de los ingresos"].fillna("Sem informação")
df["categoría económica"] = df["categoría económica"].fillna("Sem informação")


In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   código del órgano superior   1026299 non-null  int64  
 1   nombre del órgano superior   1026299 non-null  object 
 2   código del órgano            1026299 non-null  int64  
 3   nombre del órgano            1026299 non-null  object 
 4   código de la unidad gestora  1025685 non-null  float64
 5   nombre de la unidad gestora  1025685 non-null  object 
 6   categoría económica          1026299 non-null  object 
 7   origen de los ingresos       1026299 non-null  object 
 8   tipo de ingresos             1026299 non-null  object 
 9   detalle                      996962 non-null   object 
 10  valor previsto actualizado   974984 non-null   float64
 11  valor registrado             999880 non-null   float64
 12  valor ejecutado              986834 non-null   

# Eliminar columna detalle

Para nuestro análisis esta columna no nos va a proveer de información relevante 

In [41]:
df.drop(columns="detalle",inplace=True)


# Quitar valores nulos de valor previsto, registrado, ejecutado y porcentaje

Para obtener datos con sentido, haremos lo siguiente, si las columnas valor previsto, registrado, ejecutado son:
- Las tres nulas: Eliminaremos las filas, ya que no nos aportan nada de valor

El resto las tomaremos como son.

In [42]:
c = df["valor previsto actualizado"].isnull()
c1 = df["valor registrado"].isnull()
c2 = df["valor ejecutado"].isnull()
c3 = df["porcentaje ejecutado"].isnull()
df[c & c1 & c2 ].shape

(54, 15)

Tenemos 54 filas donde todo es nulo, las eliminamos para obtener mayor limpieza, además esto no afectará a los resultados

In [43]:
df.dropna(subset=["valor previsto actualizado", "valor registrado", "valor ejecutado"], how='all',inplace=True)


In [44]:
df[["porcentaje ejecutado"]] = df[["porcentaje ejecutado"]].fillna(0)
df[["valor previsto actualizado", "valor registrado", "valor ejecutado", "porcentaje ejecutado"]] = df[["valor previsto actualizado", "valor registrado", "valor ejecutado", "porcentaje ejecutado"]].astype(float)


Vemos ahora las que sean 0

In [45]:
c = df["valor previsto actualizado"] == 0
c1 = df["valor registrado"] == 0
c2 = df["valor ejecutado"] == 0 
c3 = df["porcentaje ejecutado"] == 0
df[c & c1 & c2 & c3].shape

(3428, 15)

Tenemos otras 3428 Filas donde todo es 0, lo que no aporta nada a nuestros datos, las eliminaremos también

In [46]:
# Seleccionar las filas donde todas las columnas relevantes sean 0
condicion1 = df["valor previsto actualizado"] == 0
condicion2 = df["valor registrado"] == 0
condicion3 =  df["valor ejecutado"] == 0
condicion4 = df["porcentaje ejecutado"] == 0

df.drop(df[condicion1 & condicion2 & condicion3 & condicion4].index, inplace= True)


Veamos si hay tambien momentos donde hay Nan y 0 juntos (5 pruebas más)

In [47]:
condicion1 = df["valor previsto actualizado"] == 0
condicion2 = df["valor registrado"] == 0
condicion3 =  df["valor ejecutado"].isnull()
condicion4 = df["porcentaje ejecutado"] == 0

df[condicion1 & condicion2 & condicion3 & condicion4].shape

(35509, 15)

In [48]:
df.drop(df[condicion1 & condicion2 & condicion3 & condicion4].index, inplace= True)

In [49]:
condicion1 = df["valor previsto actualizado"] == 0
condicion2 = df["valor registrado"].isnull()
condicion3 =  df["valor ejecutado"] == 0
condicion4 = df["porcentaje ejecutado"] == 0

df[condicion1 & condicion2 & condicion3 & condicion4].shape

(241, 15)

In [50]:
df.drop(df[condicion1 & condicion2 & condicion3 & condicion4].index, inplace= True)

In [51]:
condicion1 = df["valor previsto actualizado"].isnull()
condicion2 = df["valor registrado"] == 0
condicion3 =  df["valor ejecutado"] == 0
condicion4 = df["porcentaje ejecutado"] == 0

df[condicion1 & condicion2 & condicion3 & condicion4].shape

(618, 15)

In [52]:
df.drop(df[condicion1 & condicion2 & condicion3 & condicion4].index, inplace= True)

In [53]:
condicion1 = df["valor previsto actualizado"].isnull()
condicion2 = df["valor registrado"] == 0
condicion3 =  df["valor ejecutado"].isnull()
condicion4 = df["porcentaje ejecutado"] == 0

df[condicion1 & condicion2 & condicion3 & condicion4].shape

(1875, 15)

In [54]:
df.drop(df[condicion1 & condicion2 & condicion3 & condicion4].index, inplace= True)

In [55]:
condicion1 = df["valor previsto actualizado"].isnull()
condicion2 = df["valor registrado"].isnull()
condicion3 =  df["valor ejecutado"] == 0
condicion4 = df["porcentaje ejecutado"] == 0

df[condicion1 & condicion2 & condicion3 & condicion4].shape

(31, 15)

In [56]:
df.drop(df[condicion1 & condicion2 & condicion3 & condicion4].index, inplace= True)

In [57]:
condicion1 = df["valor previsto actualizado"].isnull()
condicion2 = df["valor registrado"].isnull()
condicion3 =  df["valor ejecutado"].isnull()
condicion4 = df["porcentaje ejecutado"] == 0

df[condicion1 & condicion2 & condicion3 & condicion4].shape

(0, 15)

# Ver y si fuera necesario, eliminar duplicados

Para ello:

In [58]:
df.duplicated().value_counts() 


False    967913
True      16630
Name: count, dtype: int64

In [59]:
df.drop_duplicates(keep="first",inplace=True)


In [60]:
df

código del órgano superior nombre del órgano superior  \
0                             63000   Advocacia-Geral da União   
1                             63000   Advocacia-Geral da União   
2                             63000   Advocacia-Geral da União   
3                             63000   Advocacia-Geral da União   
4                             63000   Advocacia-Geral da União   
...                             ...                        ...   
1026293                       20000   Presidência da República   
1026294                       20000   Presidência da República   
1026296                       20000   Presidência da República   
1026297                       20000   Presidência da República   
1026298                       20000   Presidência da República   

         código del órgano                                  nombre del órgano  \
0                    63000  Advocacia-Geral da União - Unidades com víncul...   
1                    63000  Advocacia-Geral da União - Unidades com víncul...   
2                    63000  Advocacia-Geral da União - Unidades com víncul...   
3                    63000  Advocacia-Geral da União - Unidades com víncul...   
4                    63000  Advocacia-Geral da União - Unidades com víncul...   
...                    ...                                                ...   
1026293              60000          Gabinete da Vice-Presidência da República   
1026294              24208     Instituto Nacional de Tecnologia da Informação   
1026296              24208     Instituto Nacional de Tecnologia da Informação   
1026297              24208     Instituto Nacional de Tecnologia da Informação   
1026298              24208     Instituto Nacional de Tecnologia da Informação   

         código de la unidad gestora  \
0                          110060.00   
1                          110060.00   
2                          110060.00   
3                          110060.00   
4                          110060.00   
...                              ...   
1026293                    110101.00   
1026294                    243001.00   
1026296                    243001.00   
1026297                    243001.00   
1026298                    243001.00   

                           nombre de la unidad gestora  categoría económica  \
0        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
1        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
2        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
3        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
4        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
...                                                ...                  ...   
1026293      GABINETE DA VICE-PRESIDENCIA DA REPUBLICA  Receitas de Capital   
1026294  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI  Receitas de Capital   
1026296  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026297  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026298  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   

            origen de los ingresos  \
0        Outras Receitas Correntes   
1        Outras Receitas Correntes   
2        Outras Receitas Correntes   
3        Outras Receitas Correntes   
4        Outras Receitas Correntes   
...                            ...   
1026293       Operações de Crédito   
1026294       Operações de Crédito   
1026296        Receita de Serviços   
1026297        Receita de Serviços   
1026298        Receita de Serviços   

                                      tipo de ingresos  \
0        Bens, Direitos e Valores Incorporados ao Patr   
1          Indenizações, restituições e ressarcimentos   
2        Multas administrativas, contratuais e judicia   
3        Bens, Direitos e Valores Incorporados ao Patr   
4          Indenizações, restituições e ressarcimentos   
...     

In [61]:
df["valor previsto actualizado"] = pd.to_numeric(df["valor previsto actualizado"], errors="coerce")
df["valor registrado"] = pd.to_numeric(df["valor registrado"], errors="coerce")
df["valor ejecutado"] = pd.to_numeric(df["valor ejecutado"],errors= "coerce")

In [63]:

c = df["porcentaje ejecutado"] != 0
e = df[c]
e

código del órgano superior  \
21                            22000   
23                            22000   
25                            22000   
29                            22000   
30                            22000   
...                             ...   
1016721                       53000   
1017108                       53000   
1020517                       44000   
1021927                       44000   
1022426                       44000   

                            nombre del órgano superior  código del órgano  \
21       Ministério da Agricultura, Pecuária e Abastec              22000   
23       Ministério da Agricultura, Pecuária e Abastec              22000   
25       Ministério da Agricultura, Pecuária e Abastec              22202   
29       Ministério da Agricultura, Pecuária e Abastec              22201   
30       Ministério da Agricultura, Pecuária e Abastec              22202   
...                                                ...                ...   
1016721         Ministério do Desenvolvimento Regional              22204   
1017108         Ministério do Desenvolvimento Regional              44205   
1020517                    Ministério do Meio Ambiente              20701   
1021927                    Ministério do Meio Ambiente              44207   
1022426                    Ministério do Meio Ambiente              44207   

                                         nombre del órgano  \
21       Ministério da Agricultura, Pecuária e Abasteci...   
23       Ministério da Agricultura, Pecuária e Abasteci...   
25             Empresa Brasileira de Pesquisa Agropecuária   
29       Instituto Nacional de Colonização e Reforma Ag...   
30             Empresa Brasileira de Pesquisa Agropecuária   
...                                                    ...   
1016721     Departamento Nacional de Obras Contra as Secas   
1017108                          Agência Nacional de Águas   
1020517  Instituto Brasileiro do Meio Ambiente e dos Re...   
1021927  Instituto Chico Mendes de Conservação da Biodi...   
1022426  Instituto Chico Mendes de Conservação da Biodi...   

         código de la unidad gestora  \
21                         130101.00   
23                         130101.00   
25                         135037.00   
29                         373001.00   
30                         135037.00   
...                              ...   
1016721                    113802.00   
1017108                    443003.00   
1020517                    193034.00   
1021927                    443032.00   
1022426                    443032.00   

                           nombre de la unidad gestora  categoría económica  \
21           COORD.-GERAL DE ORCAMENTO E FINANCAS-MAPA  Receitas de Capital   
23           COORD.-GERAL DE ORCAMENTO E FINANCAS-MAPA   Receitas Correntes   
25                                    EMBRAPA-SETORIAL   Receitas Correntes   
29       DPTO. DE ADMINISTRACAO FINANCEIRA-DAF - INCRA  Receitas de Capital   
30                                    EMBRAPA-SETORIAL   Receitas Correntes   
...                                                ...                  ...   
1016721  DEPARTAMENTO NAC. DE OBRAS CONTRA AS SECAS/UO   Receitas Correntes   
1017108           COORDENACAO DE ORCAMENTO E FINANCAS.   Receitas Correntes   
1020517   IBAMA - INST.BRASILEIRO MEIO AMBIENTE/MATRIZ   Receitas Correntes   
1021927  INST.CHICO MENDES DE CONSER.DA BIODIVERSIDADE   Receitas Correntes   
1022426  INST.CHICO MENDES DE CONSER.DA BIODIVERSIDADE   Receitas Correntes   

            origen de los ingresos  \
21               Alienação de Bens   
23            Receita Agropecuária   
25            Receita Agropecuária   
29               Alienação de Bens   
30             Receita de Serviços   
...                            ...   
1016721       Receita Agropecuária   
1017108  Outras Receitas Correntes   
1020517        Receita de Serviços   
1021927        Receita de Serviços   
1022426        Receita

   valor previsto actualizado  valor ejecutado  percentual realizado
0                        0.00          1297.13                  0.00
1                        0.00      26666621.42                  0.00
2                        0.00        301251.13                  0.00
3                        0.00          1855.58                  0.00
4                        0.00         52140.68                  0.00


código del órgano superior nombre del órgano superior  \
0                             63000   Advocacia-Geral da União   
1                             63000   Advocacia-Geral da União   
2                             63000   Advocacia-Geral da União   
3                             63000   Advocacia-Geral da União   
4                             63000   Advocacia-Geral da União   
...                             ...                        ...   
1026293                       20000   Presidência da República   
1026294                       20000   Presidência da República   
1026296                       20000   Presidência da República   
1026297                       20000   Presidência da República   
1026298                       20000   Presidência da República   

         código del órgano                                  nombre del órgano  \
0                    63000  Advocacia-Geral da União - Unidades com víncul...   
1                    63000  Advocacia-Geral da União - Unidades com víncul...   
2                    63000  Advocacia-Geral da União - Unidades com víncul...   
3                    63000  Advocacia-Geral da União - Unidades com víncul...   
4                    63000  Advocacia-Geral da União - Unidades com víncul...   
...                    ...                                                ...   
1026293              60000          Gabinete da Vice-Presidência da República   
1026294              24208     Instituto Nacional de Tecnologia da Informação   
1026296              24208     Instituto Nacional de Tecnologia da Informação   
1026297              24208     Instituto Nacional de Tecnologia da Informação   
1026298              24208     Instituto Nacional de Tecnologia da Informação   

         código de la unidad gestora  \
0                          110060.00   
1                          110060.00   
2                          110060.00   
3                          110060.00   
4                          110060.00   
...                              ...   
1026293                    110101.00   
1026294                    243001.00   
1026296                    243001.00   
1026297                    243001.00   
1026298                    243001.00   

                           nombre de la unidad gestora  categoría económica  \
0        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
1        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
2        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
3        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
4        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
...                                                ...                  ...   
1026293      GABINETE DA VICE-PRESIDENCIA DA REPUBLICA  Receitas de Capital   
1026294  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI  Receitas de Capital   
1026296  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026297  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026298  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   

            origen de los ingresos  \
0        Outras Receitas Correntes   
1        Outras Receitas Correntes   
2        Outras Receitas Correntes   
3        Outras Receitas Correntes   
4        Outras Receitas Correntes   
...                            ...   
1026293       Operações de Crédito   
1026294       Operações de Crédito   
1026296        Receita de Serviços   
1026297        Receita de Serviços   
1026298        Receita de Serviços   

                                      tipo de ingresos  \
0        Bens, Direitos e Valores Incorporados ao Patr   
1          Indenizações, restituições e ressarcimentos   
2        Multas administrativas, contratuais e judicia   
3        Bens, Direitos e Valores Incorporados ao Patr   
4          Indenizações, restituições e ressarcimentos   
...     